# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

导入数据分析所需要的库,并通过pandas的"read_csv"函数,将原始数据文件"e_commerce.csv"里的数据内容,解析为DataFrame,并赋值给变量"original_data"。

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv("e_commerce.csv")

In [3]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

在这一部分,我将对上一b部分建立的"original_data"这个DataFrame所包含的数据进行评估。

评估主要从两个方面进行:结构和内容,即整齐度和干净度。
数据的结构性问题指不符合"每列是一个变量,每行是一个观察值,每个单元格是一个值"这三个标准,数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

In [4]:
original_data.sample(10)

InvoiceNo StockCode                          Description  Quantity  \
494058    578260     23311          VINTAGE CHRISTMAS STOCKING          6   
73110     542303    35810A         ENAMEL PINK COFFEE CONTAINER         6   
346386    567184     23013        GLASS APOTHECARY BOTTLE TONIC         1   
329693    565850     22173   METAL 4 HOOK HANGER FRENCH CHATEAU         3   
323380    565295     47580                TEA TIME DES TEA COSY         6   
120063    546646     20981        12 PENCILS TALL TUBE WOODLAND        24   
89931     543979     84880                WHITE WIRE EGG HOLDER         1   
467426    576339    85170D      SET/6 PINK BIRD T-LIGHT CANDLES         2   
34056     539317    85123A   WHITE HANGING HEART T-LIGHT HOLDER         4   
532348    580983     22577  WOODEN HEART CHRISTMAS SCANDINAVIAN        16   

             InvoiceDate  UnitPrice  CustomerID         Country  
494058  11/23/2011 13:15       2.55     17469.0  United Kingdom  
73110    1/27/2011 11:05       2.10     12539.0           Spain  
346386   9/18/2011 15:41       3.95     16033.0  United Kingdom  
329693    9/7/2011 12:55       3.29     14530.0  United Kingdom  
323380    9/2/2011 12:01       2.55     12794.0        Portugal  
120063   3/15/2011 13:49       0.85     13716.0  United Kingdom  
89931    2/14/2011 15:43       7.50     14606.0  United Kingdom  
467426  11/14/2011 15:27       4.13     14096.0  United Kingdom  
34056   12/16/2010 19:04       2.95     17392.0  United Kingdom  
532348   12/6/2011 16:26       0.83         NaN  United Kingdom

从抽样的10行来看都符合"每列是一个变量,每行是一个观察值,每个单元格是一个值"这三个标准

## 清理数据

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出的结果来看,数据共有541909条观察值, 而Description和CustomerID 变量存在缺失值。

此外,InvoiceDate的数据类型应为日期, CustomerID的数据类型应为字符串,应当进行数据格式转换。

### 评估缺失数据

在了解DataFrame存在缺失值后，根据条件提取出缺失观察值。

In [6]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

有1454条数据缺失Description变量值
从输出的结果来看，这些缺失Description的交易数据，UnitPrice都为0，为了验证猜想，我们加筛选条件，
看是否存在Description变量缺失且UnitPrice不为0的数据

In [7]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

筛选出的结果数量为0条，说明缺失Description值的数据，同时也不具备有效的UnitPrice值。

Description表示产品的名称，UnitPrice表示产品的单价，都是进行后续商品交易分析的重要变量。

如果它们同时缺失/无效，我们认为数据无法提供有效含义，因此在些后续可以被删除。

In [8]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

CustomerID表示客户编号，不是分析畅销商品的必要变量。

并且从输出结果来看，有些CustomerID缺失的交易数据仍然有效，因此保留些变量为为空的观察值。

### 评估重复数据

根据数据变量的含以来看，虽然InvoiceNo、StockCode和CustomerID都是唯一标识，但一次交易可能包含多件商品，因此InvoiceNo可以存在重复，

顾客可以进行多次交易或下单多个商品，因此CustomerID也可以存在重复。

那么针对些数据集，我们无续评估重复数据。

### 评估不一致数据

不一致数据可能存在于Country变量中，我们要查看是否存在多个不同值指代同一国家的情况。

In [9]:
original_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


从Country变量来看，"USA"、"United states"均在表示美国，"United Kingdom"、"UK"、"U.K"均在表示英国，因此应该在清洗步骤对这些值进行统一，只保留一个指代值。

评估无效或错误数据

可以通过DataFarme的describe方法，对数值统计信息进行快速了解。

In [10]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看，Quantity和UnitPrice存在负数，会对后续数值分析产生影响。

因此，我们先筛选出Quantity数值为负数的观察值，进一步评估其含义。

In [11]:
original_data[original_data["Quantity"] < 0 ]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

从筛选结果来看，Quantity变量为负数的观察值，InvoiceNo似乎都是以c开头的，表示订单取消。

但此猜测不一定准确，为了验证猜想，我们增加筛选条件，看是否存在Quantity变量为负数且InvoiceNo不以c开头的观察值。

In [12]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

以上猜想被证实错误，因为还存在Quantity变量为负数且InvoiceNo不以c开头的观察值。

但根据以上输出结果，这些筛选出的观察值的UnitPrice均为0，因此增加UnitPrice的条件进行验证。

In [13]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C") & (original_data["UnitPrice"]!= 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到证实，当Quantity变量为负数的时候，观察值满足以下条件之一:

InvoiceNo以C开头，表示订单取消  
UnitPrice为0，说明单价为0英镑

这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清理步骤，将Quantity为负数的观察值删除。

In [15]:
original_data[original_data["UnitPrice"]<0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

从输出结果来看，	UnitPrice为负数的观察值都是坏帐调账，不属于实际商品交易数据，因此为在数据清洗步骤中也将其删除。

### 清洗数据

根据前面评估部分得到的结论，我们需要进行的数据清理包括:

把InvoiceDate变量的数据类型转换为日期时间  
把CustomerID变量的数据类型转换为字符串  
把Description变量缺失的观察值删除  
把Country变量值"USA"替换为"United States"  
把Country变量值"UK"、"U.K."替换为"United Kingdom"  
把Quantity变量值为负数观察值删除  
把UnitPrice变量值为负数观察值删除

为了区分开经过清理的数据和原始的数据，我们创建新的变量cleaned_data，让它为original_data复制出的副本。  
我们之后的清理步骤都将被运用在cleaned_data上。

In [16]:
cleaned_data = original_data.copy()
cleaned_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

把InvoiceDate变量的数据类型转换为日期时间类型:

In [19]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把CustomerID变量的数据类型转换为字符串:

In [21]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把CustomerID变量值结尾的".0"删除:

In [23]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0,-2)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把Description变量缺失的观察值删除，并查看删除后该列空缺值个数和:

In [28]:
cleaned_data.dropna(subset=["Description"],inplace=True)

In [29]:
cleaned_data["Description"].isnull().sum()

0

把Country变量"USA"替换为"United States"，并检查替换后"USA"变量个数:

In [30]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA":"United States"})

In [31]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

把Countyry变量值"UK"、"U.K."替换为"United Kingdom"，并检查替换后"UK" 和 "U.K."变量个数:

In [32]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK":"United Kingdom","U.K.":"United Kingdom"})

In [34]:
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K."]))

0
0


把Quantity变量值为负数的观察值删除，并检查替换后Quantity变量为负数的个数:

In [36]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >=0]

In [39]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

把UnitPrice变量值为负数的观察值删除，并检查替换后UnitPrice变量为负数的个数:

In [40]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >=0]

In [41]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## 保存清理后的数据

完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为e_commerce_cleaned.csv。

In [43]:
cleaned_data.to_csv("e_commerce_cleaned.csv",index=False)

In [44]:
pd.read_csv("e_commerce_cleaned.csv").head()

C:\Users\陈\AppData\Local\Temp\ipykernel_16352\2686214642.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom